File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    public String reverseWords(String s) {
        char[] ch = s.toCharArray();
        char[] res = new char[ch.length + 1];
        
        int i = ch.length - 1, m = 0;
        while (i >= 0) {
            while (i >= 0 && ch[i] == ' ') {
                i--;
            }
            if (i < 0) break;
            int j = i;
            while (i >= 0 && ch[i] != ' ') {
                i--;
            }
            for (int k = i + 1; k <= j; k++) {
                res[m++] = ch[k];
            }
            res[m++] = ' ';
        }
        return new String(res, 0, m - 1);
    }
}

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        string res;
        
        for (int i = s.size() - 1; i >= 0;) {
            while (i >= 0 && s[i] == ' ') {
                i--;
            }
            
            if (i < 0) {
                break;
            }
            
            int j = i;
            while (i >= 0 && s[i] != ' ') {
                i--;
            }
            
            for (int k = i + 1; k <= j; k++) {
                res = res + s[k];
            }
            
            res = res + ' ';
        }
        
        res.pop_back();
        return res;
    }
};

static const auto __optimize__ = []() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
    return 0;
}();